onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -155,6 +159,46 @@ CUDAExecutionProvider::PerThreadContext::~PerThreadContext() {
   }
 }
 
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 10000
+bool CUDAExecutionProvider::PerThreadContext::IsGraphCaptureAllowed() const {
+  return regular_run_count_before_graph_capture_ >= min_num_runs_before_cuda_graph_capture_;
+}
+
+void CUDAExecutionProvider::PerThreadContext::CaptureBegin()  {
+  cuda_graph_.Reset();
+  cuda_graph_.CaptureBegin();
+}
+
+void CUDAExecutionProvider::PerThreadContext::CaptureEnd() {
+  cuda_graph_.CaptureEnd();
+  is_graph_captured_ = true;
+}
+
+bool CUDAExecutionProvider::PerThreadContext::IsGraphCaptured() const {
+  return is_graph_captured_;
+}
+
+Status CUDAExecutionProvider::PerThreadContext::ReplayGraph() {
+  if (!IsGraphCaptureAllowed()) {
```

Comment:
Nit: Will just `if (!IsGraphCaptured()) { return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Cuda graph is not captured yet.}` do here instead of 2 checks (also IsGraphCaptureAllowed() doesn't really seem relevant in a method that is focussed on doing a graph replay) returning a message not really actionable by the user ? The user only invokes `InferenceSession::Run()` and our internal logic is the one that decides when to invoke `ReplayGraph()`. If we do have a bug there, I feel it is enough to alert the user that the graph has not been captured yet instead of fiving details like "run n regular runs..." which might confuse them.

Reply:
The checks and error messages are more for ORT backend developers to understand the root cause if any issues come up when replaying graph. I agree the message may make users confused, and `ORT_ENFORCE(IsGraphCaptured())` may be enough for developers. I have changed it as suggested.